src/core/sign.ml

```diff
@@ -273,7 +273,9 @@ let read : string -> t = fun fname ->
     its dependencies. *)
 let add_rule : t -> sym -> rule -> unit = fun sign sym r ->
   sym.sym_rules := !(sym.sym_rules) @ [r];
-  if sym.sym_path <> sign.sign_path then
+  if sym.sym_path <> sign.sign_path &&
+     not (sym == Unif.Hint.atom || sym == Unif.Hint.list)
+  then
```

Comment:
I think this change should not be necessary if things are done correctly with your special symbols. I think it is better to give them impossible paths (e.g., the empty path), and not to rely on specific code here.